File: sudocode-go/internal/sudocodefile/sudocodefile.go
// Package sudocodefile handles the .sudocode file/dir indirection,
// following the same pattern as Git's gitdir: mechanism.
package sudocodefile

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

const (
	// FileName is the name of the .sudocode file or directory.
	FileName = ".sudocode"
	// Prefix is the required prefix in a .sudocode file.
	Prefix = "sudocodedir: "
)

// ErrNotFound indicates no .sudocode file or directory exists.
var ErrNotFound = errors.New("sudocodefile: .sudocode not found")

// ErrMalformed indicates the .sudocode file has invalid content.
var ErrMalformed = errors.New("sudocodefile: malformed .sudocode file")

// ResolveSudocodeDir resolves the sudocode data directory for a repo.
//
// If .sudocode is a directory, it returns its path directly.
// If .sudocode is a file, it reads the sudocodedir: line and resolves the path.
// Relative paths are resolved relative to repoPath.
func ResolveSudocodeDir(repoPath string) (string, error) {
	p := filepath.Join(repoPath, FileName)

	info, err := os.Stat(p)
	if err != nil {
		if os.IsNotExist(err) {
			return "", ErrNotFound
		}
		return "", fmt.Errorf("sudocodefile: stat: %w", err)
	}

	if info.IsDir() {
		return p, nil
	}

	// It's a file — read and parse
	data, err := os.ReadFile(p)
	if err != nil {
		return "", fmt.Errorf("sudocodefile: read: %w", err)
	}

	content := strings.TrimRight(string(data), "\n\r")
	if !strings.HasPrefix(content, Prefix) {
		return "", fmt.Errorf("%w: missing %q prefix", ErrMalformed, Prefix)
	}

	dir := strings.TrimPrefix(content, Prefix)
	if dir == "" {
		return "", fmt.Errorf("%w: empty path after prefix", ErrMalformed)
	}

	if !filepath.IsAbs(dir) {
		dir = filepath.Join(repoPath, dir)
	}

	return filepath.Clean(dir), nil
}

// WriteSudocodeFile writes a .sudocode file with a sudocodedir: line
// pointing to the given sudocode data directory.
func WriteSudocodeFile(repoPath, sudocodeDir string) error {
	p := filepath.Join(repoPath, FileName)
	content := Prefix + sudocodeDir + "\n"
	return os.WriteFile(p, []byte(content), 0o644)
}

File: cli/src/sudocode-file.ts
/**
 * .sudocode file/dir indirection — follows Git's gitdir: pattern.
 *
 * When .sudocode is a directory: data lives there (co-located mode).
 * When .sudocode is a file: contains a single line `sudocodedir: <path>` pointing to the data directory.
 */

import * as fs from "fs";
import * as path from "path";

const FILE_NAME = ".sudocode";
const PREFIX = "sudocodedir: ";

/**
 * Resolve the sudocode data directory for a repository.
 *
 * - If .sudocode is a directory, returns its absolute path.
 * - If .sudocode is a file, reads the `sudocodedir:` line and resolves the path.
 * - If .sudocode does not exist, returns null.
 * - If .sudocode file is malformed, throws an error.
 *
 * Relative paths in the file are resolved relative to repoPath.
 */
export function resolveSudocodeDir(repoPath: string): string | null {
  const p = path.join(repoPath, FILE_NAME);

  let stat: fs.Stats;
  try {
    stat = fs.statSync(p);
  } catch {
    return null;
  }

  if (stat.isDirectory()) {
    return p;
  }

  // It's a file — read and parse
  const content = fs.readFileSync(p, "utf8").replace(/[\r\n]+$/, "");

  if (!content.startsWith(PREFIX)) {
    throw new Error(
      `Malformed .sudocode file: missing "${PREFIX}" prefix`
    );
  }

  const dir = content.slice(PREFIX.length);
  if (!dir) {
    throw new Error("Malformed .sudocode file: empty path after prefix");
  }

  if (path.isAbsolute(dir)) {
    return path.normalize(dir);
  }

  return path.resolve(repoPath, dir);
}

/**
 * Write a .sudocode file with a `sudocodedir:` line pointing to the given directory.
 */
export function writeSudocodeFile(
  repoPath: string,
  sudocodeDir: string
): void {
  const p = path.join(repoPath, FILE_NAME);
  fs.writeFileSync(p, `${PREFIX}${sudocodeDir}\n`, "utf8");
}
